daemon: Create all global maps in cilium-agent

pchaigno opened the pull request:
The first 6 commits are preparatory work for the last 9 commits, which creates all global maps in the cilium-agent. The last 9 commits can be easily squashed if preferred as they're fairly similar (I initially broke them down to bisect a bug).

Fixes: #9276

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10241)
<!-- Reviewable:end -->

tklauser commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -249,19 +259,20 @@ func (m *Map) DumpEntries() (string, error) {
 
 // NewMap creates a new CT map of the specified type with the specified name.
 func NewMap(mapName string, mapType MapType) *Map {
```
While at it, I think you could unexport this function (maybe in a separate commit). It is only used within the `ctmap` package.

pchaigno replied:
Done.

pchaigno committed:
maps/metricsmap, common: Move getNumPossibleCPUs to common.util

getNumPossibleCPUs is needed for all per-cpu map and perf event arrays.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Create method to create BPF maps

Contrary to OpenOrCreate, Create doesn't keep the BPF map opened. It is
used to create all global maps from the daemon.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/lbmap, bpf: Declare rev. NAT sock map constants on Go side

This commit declares the size and names of reverse NAT sock maps as
constants on the Go side and writes them into the BPF program when
generated.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
alignchecker: Support checking types' sizes only

and use it to check that all Go-side keys and values for maps have the
correct sizes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/policymap: Rename policy call map to clarify intent

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/ctmap: Create NAT maps with correct BPF map type

NAT maps were previously always created as type MapTypeLRUHash.  However,
to use these Go objects to create the BPF maps on the kernel side, we
need the Go and BPF map types to be the same.

Because of this change, we also cannot initialize ctmap's mapInfo in
ctmap.init() and need to use a singleton (sync.Once) pattern.  We also
don't need to call ctmap.InitMapInfo from NewDaemon since that
information will be lazily initialized when it's used.  Thus,
ctmap.InitMapInfo is unexported.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/lbmap, daemon: Create SockRevNat map from daemon

Create the sock reverse NAT maps during daemon initialization.  The maps
are closed as soon as they're created as the cilium-agent never uses them.

To update the deepcopy files for the new maps, I used:

    make generate-k8s-api

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/eventsmap, daemon: Create map from daemon

Create the events map during daemon initialization.  The map is closed
as soon as it's created since cilium-agent never uses it.

To generate the deepcopy file for the new map, I used:

    make generate-k8s-api

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/signalmap, daemon: Create map from daemon

Create the signal map during daemon initialization.  The map is closed
as soon as it's created since cilium-agent never uses it.

To generate the deepcopy file for the new map, I used:

    make generate-k8s-api

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/neighborsmap, daemon: Create map from daemon

Create neighbors maps used to store IP-MAC associations when NodePort
services are implemented in BPF datapath.

To generate the deepcopy file for the new maps, I used:

    make generate-k8s-api

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/policymap, daemon: Create policy maps from daemon

Create per-endpoint policy maps during daemon initialization.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/encrypt, daemon: Create map from daemon

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/ctmap, daemon: Create maps from daemon

Create global and per-endpoint ct maps during daemon initialization.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/policymap, daemon: Create policy call map from daemon

Create the policy call map during daemon initialization.  The map is
closed as soon as it's created since cilium-agent never uses it.

To generate the deepcopy files for the new maps, I used:

    make generate-k8s-api

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
maps/nat, daemon: Create maps from daemon

Create global NAT maps during daemon initialization.

Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet commented on cilium/cmd/bpf_ct_flush.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017-2018 Authors of Cilium
+// Copyright 2017-2020 Authors of Cilium
```
This file is not changed otherwise in the PR, so let's remove copyright year change?

qmonnet replied:
Not related, but there's a typo in this commit's log: “were previously always create” -> created (but GH doesn't offer to comment on commit log it seems).

qmonnet commented on cilium/cmd/bpf_ct_list.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017-2018 Authors of Cilium
+// Copyright 2017-2020 Authors of Cilium
```
Same here, no change in file.

qmonnet commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016-2019 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
```
File not changed.

qmonnet commented on pkg/bpf/map_linux.go:
```diff
@@ -452,6 +452,16 @@ func (m *Map) OpenOrCreateUnpinned() (bool, error) {
+	if err != nil {
+		return isNew, err
+	}
+	return isNew, m.Close()
```
Commit log: “doesn't open” -> “doesn't keep open”?

qmonnet commented on pkg/maps/neighborsmap/neighborsmap.go:
```diff
@@ -0,0 +1,115 @@
+		neighbors6Map := bpf.NewMap(Map6Name,
+			bpf.MapTypeLRUHash,
+			&Key6{},
+			int(unsafe.Sizeof(Key4{})),
```
Should this be `Key6{}` instead?
